### DISCUSSION TITLE
How to use a button in a Page to trigger a function in implementation::MainWindow class?

### DISCUSSION THREAD

**Discussion by Extremeer** [microsoft/WindowsAppSDK#3167] [Q&A] [answered]
There's a funtion MainWindow::GetHWND in implementation::MainWindow.xmal.cpp. I want to use a button in HomePage.xmal to trigger the function to get the handle of MainWindow. But if the code is:
```
//in MainWindow.xaml.cpp
HWND MainWindow::GetHWND()
{
    winrt::ProjectName::MainWindow thisWindow = *this;
    auto windowNative{ thisWindow.as<::IWindowNative>() };
    HWND hWnd{ 0 };
    return hWnd;
}
```

Method 1 occurred an error: A non static member reference must be relative to a specific object.
```
//in Page.xmal.cpp
void Page::myButton_Click(IInspectable const&, RoutedEventArgs const&)
{
    HWND = MainWindow::GetHWND();
}
```

If I'm trying to instantiate the MainWindow class, method 2 occurred another error: Objects of abstract class type are not allowed.
```
//in Page.xmal.cpp
void Page::myButton_Click(IInspectable const&, RoutedEventArgs const&)
{
    mainw = new MainWindow;
    HWND = mainw.GetHWND();
}
```

**Comment by Extremeer** [comment, marked as answer]
Use a global flag solve the problem.

**Comment by DarranRowe** [comment]
The thing to remember is that the default application class stores an instance of the main window. It stores this when the application is launched, during the OnLaunched handler. What this means is that it is always possible to use this to get an instance of the main window.

By default, you should see the App class is similar to:

```
namespace winrt::App1::implementation
{
	struct App : AppT<App>
	{
		App();

		void OnLaunched(Microsoft::UI::Xaml::LaunchActivatedEventArgs const&);

	private:
		winrt::Microsoft::UI::Xaml::Window window{ nullptr };
	};
}
```

There are namespace differences, but this is the basic implementation. There is nothing stopping you adding a member function to retrieve the window instance.

```
namespace winrt::App1::implementation
{
	struct App : AppT<App>
	{
		App();

		void OnLaunched(Microsoft::UI::Xaml::LaunchActivatedEventArgs const&);

		winrt::Microsoft::UI::Xaml::Window Window() const;

	private:
		winrt::Microsoft::UI::Xaml::Window window{ nullptr };
	};
}
```

This would just simply return window. You can then get your application instance using Application.Current:

```
auto current_generic_application = winrt::Microsoft::UI::Xaml::Application::Current();
//Instance of the app class as the proper type, note that this is of type ComRef<App>.
//Meaning that you need to use -> to access member functions.
auto current_application = current_generic_application.as<winrt::App1::implementation::App>();
```

From there, you can use as to get your main window and get the HWND.

**Comment by Extremeer** [comment]
Thank you so much! Nice method for the question!
